File: multiply.rs
fn mul(mat1: &[i32], mat2: &[i32], column_result: &usize) -> Vec<i32> {
// fn mul(weights_n: &[i32], outputs_n_minus_1: &[i32], default_value_one: &usize) -> Vec<i32> {

    // the result is stored here
    let mut mat_result: Vec<i32> = Vec::new();

    let number_row_mat2: usize = mat2.len() / column_result;
    let number_row_mat1: usize = mat1.len() / number_row_mat2;
        
    for row_mat1 in 0..number_row_mat1 {
    // for each row of mat1

        // The number of columns of the second matrix (mat2) is  
        // equal to the number of columns of the matrix as the result :
        for col_mat2 in 0..*column_result {
        // for each collumn of mat2
            
            let mut result: i32 = 0;
    
            for row_mat2 in 0..number_row_mat2 {
            // for each row of mat2

                // The number columns of mat1 is equal
                // to the number of rows of mat2 :
                let index_mat1: usize = (number_row_mat2 * row_mat1) + row_mat2;

                // The number of columns of the second matrix (mat2) is equal 
                // to the number of columns of the matrix as the result :
                let index_mat2: usize = (*column_result * row_mat2) + col_mat2;

                let pre_result: i32 = mat1[index_mat1] * mat2[index_mat2];
                result += pre_result;
            }
    
            mat_result.push(result);
        }
    }
    
    mat_result
    // neuron outputs n
}


fn main() {
    println!("\nlayer 1 :");
    let weights_1: Vec<i32> = vec![0, 1, 2, 3, 4, 5];
    let inputs: Vec<i32> = vec![0, 1, 2];
    // the matrix as the result must have 3 columns
    let column_result: usize = 1; // we want to get vector, so with only 1 column
    let neurones_1: Vec<i32> = mul(&weights_1, &inputs, &column_result);
    println!("\nNumber at the neurones of the layer 1 : {:?}", neurones_1);


    println!("\nlayer 2 :");
    let weights_2: Vec<i32> = vec![0, 1, 2, 3, 4, 5];
    let column_result: usize = 1; // we want to get vector, so with only 1 column
    let neurones_2: Vec<i32> = mul(&weights_2, &neurones_1, &column_result);
    println!("\nNumber at the neurones of the layer 2 : {:?}", neurones_2);


    println!("\nTest 3 :");
    let mat1: Vec<i32> = vec![0, 1, 2, 3];
    let mat2: Vec<i32> = vec![0, 1, 2, 3, 4, 5];
    let column_result: usize = 3;
    let result3: Vec<i32> = mul(&mat1, &mat2, &column_result);
    println!("\nresult3 : {:?}", result3);


    println!("\nTest 4 :");
    let mat1: Vec<i32> = vec![0, 1, 2, 3, 4, 5];
    let mat2: Vec<i32> = vec![0, 1, 2, 3];
    let column_result: usize = 2;
    let result4: Vec<i32> = mul(&mat1, &mat2, &column_result);
    println!("\nresult4 : {:?}", result4);
}

File: multiply.c
#include <stdio.h>
#include <stdlib.h>

//gcc multiply.c && valgrind./a.out

int* mul(const int *mat1, const int *mat2, const int *column_result, const size_t size1, const size_t size2) {
    int number_row_mat2 = size2 / (*column_result);

    int number_row_mat1 = size1 / number_row_mat2;

    // the number of rows of mat1 is equal to the number of 
    // rows of the matrix as the result
    int *mat_result = (int*)malloc(((*column_result) * number_row_mat1) * sizeof(int));
    //printf("\nle resultat doit contenir %d nombres", sizeof(*mat_result));

    for (int row_mat1 = 0; row_mat1 < number_row_mat1; row_mat1++) {

        // The number of columns of the second matrix (mat2) is  
        // equal to the number of columns of the matrix as the result :
        for (int col_mat2 = 0; col_mat2 < *column_result; col_mat2++) {

            int result = 0;

            for (int row_mat2 = 0; row_mat2 < number_row_mat2; row_mat2++) {

                // The number columns of mat1 is equal
                // to the number of rows of mat2 :
                int index_mat1 = (number_row_mat2 * row_mat1) + row_mat2;

                // The number of columns of the second matrix (mat2) is equal 
                // to the number of columns of the matrix as the result :                
                int index_mat2 = (*column_result * row_mat2) + col_mat2;

                int pre_result = mat1[index_mat1] * mat2[index_mat2];
                result += pre_result;
            }
            
            // The number of columns of the second matrix (mat2) is equal 
            // to the number of columns of the matrix as the result : 

            // the number of rows of mat1 is equal to the number of 
            // rows of the matrix as the result
            mat_result[(*column_result * row_mat1) + col_mat2] = result;
        }
    }

    return mat_result;
}

int main() {

    printf("\nlayer 1 :\n");
    int weights_1[] = {0, 1, 2, 3, 4, 5};
    int inputs[] = {0, 1, 2};

    int column_result = 1; // 1 because we want a vector

    size_t size1 = sizeof(weights_1) / sizeof(weights_1[0]);
    size_t size2 = sizeof(inputs) / sizeof(inputs[0]);
    int *neurones_1 = mul(weights_1, inputs, &column_result, size1, size2);
    //int *neurones_1 = mul(weights_1, inputs, &column_result);

    int iteratore_layer_1 = size1 / size2;
    printf("\nNumber at the neurones of the layer 1 : ");
    for (int i = 0; i < iteratore_layer_1; i++) {
        printf("%d ", neurones_1[i]);
    }
    printf("\n");

    printf("\nlayer 2 :\n");
    int weights_2[] = {0, 1, 2, 3, 4, 5};
    size_t size3 = sizeof(weights_2) / sizeof(weights_2[0]);
    size_t size4 = sizeof(neurones_1) / sizeof(neurones_1[0]);
    int *neurones_2 = mul(weights_2, neurones_1, &column_result, size3, size4);
    //int *neurones_2 = mul(weights_2, neurones_1, &column_result);

    int iteratore_layer_2 = size3 / size4;
    printf("\nNumber at the neurones of the layer 2 : ");
    for (int i = 0; i < iteratore_layer_2; i++) {
        printf("%d ", neurones_2[i]);
    }
    printf("\n");

    free(neurones_1);
    free(neurones_2);

    return 0;
}
